### PR TITLE
refactor(ingest): Use shared connection wrapper over connection cache

### PR DESCRIPTION
**Pull request by asikowitz**
Rather than handle sharing sqlite connections via the filename-based `_SqliteConnectionCache`, creates a SQlite `ConnectionWrapper` that can be passed into `FileBacked` objects. Personally liked this more, but happy to discuss pros and cons.

Pros:
- Connection sharing is more transparent, via a shared object, rather than via a shared filename Path
- User doesn't need to worry about creating a ingestion-run-unique filename
- Allows for automatic file cleanup
- Removes complexity around SqliteConnectionCache

Cons:
- Adds complexity / another layer with a wrapper around the SQlite connection
  * I considered inheriting from `sqlite.Connection` directly, rather than wrapping it, but didn't want to worry about interfering with its internals
- Filenames no longer can be specified when creating `FileBacked` objects -- has to be done by creating a `ConnectionWrapper` and passing it in. I did this to reduce complexity around filenames, so they don't have to be stored in 2 places. Overall don't think this is too big a deal because I don't expect we'll need to manually specify filenames much

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/file_backed_collections.py:124]: i think we need to keep the field() stuff because this is a dataclass

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/file_backed_collections.py:299]: if they manually call close, and then the gc calls `__del__`, we'll end up trying to delete twice?

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/file_backed_collections.py:88]: same as below - if the call close manually / via the contextmanager and then the garbage collector calls `__del__`, then we'll call cleanup twice

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (asikowitz) [reply metadata-ingestion/src/datahub/utilities/file_backed_collections.py:53]: Yup I'm bad

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (asikowitz) [reply metadata-ingestion/src/datahub/utilities/file_backed_collections.py:124]: I don't think so. We do this in other places, e.g. in https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic.py. https://www.infoworld.com/article/3563878/how-to-use-python-dataclasses.html#toc-3 goes into more detail -- you can hold off using `field` for the most basic cases

Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/file_backed_collections.py:124]: Yup you're right - I misremembered it

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (asikowitz) [reply metadata-ingestion/src/datahub/utilities/file_backed_collections.py:299]: Yeah, I was aiming for an idempotent close. It's a little awkward because `flush` gets called multiple times (which is why I ended up adding the check on `_prune_cache`) so i'm not opposed to preventing close from getting called multiple times. However, close does end up getting called multiple times in all the tests and it ends up fine, so I don't think it's required

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (asikowitz) [reply metadata-ingestion/src/datahub/utilities/file_backed_collections.py:53]: Actually, it seems like this is acceptable:
https://peps.python.org/pep-0526/#class-and-instance-variable-annotations
https://docs.python.org/3/library/typing.html#typing.ClassVar

Which makes some sense because it'd be weird if adding a default to a type annotation converts it from an instance var to a class var. Functionally, this means you can change the default value of `_directory` in code by setting `ConnectionWrapper._directory = <some_value>`, which you couldn't if we just set this in `__init__`, but I guess we just shouldn't do that...

Overall don't really like this syntax with python typing, especially since mypy doesn't error if you reassign `ConnectionWrapper._directory`

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/file_backed_collections.py]: oh i see - because the cleanup() call is idempotent, it should be fine

File: metadata-ingestion/src/datahub/utilities/file_backed_collections.py
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/file_backed_collections.py]: it might be appropriate to override the init then, but I'll leave that up to you

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**
